codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java

```diff
@@ -16,46 +16,58 @@
 
 package io.netty.handler.codec.http2;
 
+import io.netty.buffer.ByteBufAllocator;
 import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
 import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
 import io.netty.handler.codec.MessageToMessageCodec;
 import io.netty.handler.codec.http.DefaultHttpContent;
 import io.netty.handler.codec.http.DefaultLastHttpContent;
-import io.netty.handler.codec.http.FullHttpRequest;
-import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.FullHttpMessage;
 import io.netty.handler.codec.http.HttpContent;
 import io.netty.handler.codec.http.HttpHeaderNames;
 import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMessage;
 import io.netty.handler.codec.http.HttpObject;
 import io.netty.handler.codec.http.HttpRequest;
-import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpScheme;
 import io.netty.handler.codec.http.HttpUtil;
 import io.netty.handler.codec.http.HttpVersion;
 import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.ssl.SniHandler;
+import io.netty.handler.ssl.SslHandler;
 import io.netty.util.ReferenceCountUtil;
 import io.netty.util.internal.UnstableApi;
 
 import java.util.List;
 
 /**
- * This is a server-side adapter so that an http2 codec can be downgraded to
- * appear as if it's speaking http/1.1.
+ * This handler converts from {@link Http2StreamFrame} to {@link HttpObject},
+ * and back. It can be used as an adapter in conjunction with {@link
+ * Http2MultiplexCodec} to make http/2 connections backward-compatible with
+ * {@link ChannelHandler}s expecting {@link HttpObject}
  *
- * In particular, this handler converts from {@link Http2StreamFrame} to {@link
- * HttpObject}, and back.  For simplicity, it converts to chunked encoding
- * unless the entire stream is a single header.
+ * For simplicity, it converts to chunked encoding unless the entire stream
+ * is a single header.
  */
 @UnstableApi
-public class Http2ServerDowngrader extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
-
+public class Http2StreamFrameToHttpObjectCodec extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
+    private final boolean isServer;
     private final boolean validateHeaders;
 
-    public Http2ServerDowngrader(boolean validateHeaders) {
+    private volatile HttpScheme scheme;
+
+    public Http2StreamFrameToHttpObjectCodec(final boolean isServer,
```

Comment:
@lionelli1 nit: either make both final or none of these :)

Reply:
oops, missed that one out. done!